src/tls.rs

```diff
@@ -1,6 +1,354 @@
+                keep_alive_count: kac.count as i32,
+            };
+            rcfg.keep_alive_cfg = &mut raw_kac as *mut _;
+        }
+
+        let mut raw_psk: esp_idf_sys::psk_key_hint;
+        if let Some(psk) = &cfg.psk_hint_key {
+            raw_psk = esp_idf_sys::psk_key_hint {
+                key: psk.key.as_ptr(),
+                key_size: psk.key.len(),
+                hint: psk.hint.as_ptr(),
+            };
+            rcfg.psk_hint_key = &mut raw_psk as *mut _;
+        }
+
+        if cfg.use_crt_bundle_attach {
+            rcfg.crt_bundle_attach = Some(esp_idf_sys::esp_crt_bundle_attach);
+        }
+
+        rcfg.is_plain_tcp = cfg.is_plain_tcp;
+        rcfg.if_name = std::ptr::null_mut();
+
+        let tls = unsafe { esp_idf_sys::esp_tls_init() };
+        if tls == std::ptr::null_mut() {
+            return Err(EspError::from_infallible::<ESP_ERR_NO_MEM>());
+        }
+        let ret = unsafe {
+            esp_idf_sys::esp_tls_conn_new_sync(
+                host.as_bytes().as_ptr() as *const i8,
+                host.len() as i32,
+                port as i32,
+                &rcfg,
+                tls,
+            )
+        };
+
+        if ret == 1 {
+            Ok(EspTls {
+                reader: EspTlsRead { raw: tls },
+                writer: EspTlsWrite { raw: tls },
+            })
+        } else {
+            unsafe {
+                esp_idf_sys::esp_tls_conn_destroy(tls);
+            }
+
+            Err(EspError::from_infallible::<ESP_FAIL>())
+        }
+    }
+
+    pub fn read(&mut self, buf: &mut [u8]) -> Result<usize, EspIOError> {
+        self.reader.read(buf)
+    }
+
+    pub fn write(&mut self, buf: &[u8]) -> Result<usize, EspIOError> {
+        self.writer.write(buf)
+    }
+
+    pub fn split(&mut self) -> (&mut EspTlsRead, &mut EspTlsWrite) {
+        (&mut self.reader, &mut self.writer)
+    }
+}
+
+impl io::Io for EspTls {
+    type Error = EspIOError;
+}
+
+impl io::Read for EspTls {
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize, EspIOError> {
+        self.read(buf)
+    }
+}
+
+impl io::Write for EspTls {
+    fn write(&mut self, buf: &[u8]) -> Result<usize, EspIOError> {
+        self.write(buf)
+    }
+
+    fn flush(&mut self) -> Result<(), EspIOError> {
+        Ok(())
+    }
+}
+
+pub struct EspTlsRead {
+    raw: *mut esp_idf_sys::esp_tls,
+}
+
+impl EspTlsRead {
+    pub fn read(&mut self, buf: &mut [u8]) -> Result<usize, EspIOError> {
+        // cannot call esp_tls_conn_read bc it's inline
+        //let ret = unsafe { esp_idf_sys::esp_tls_conn_read(self.raw, buf.as_mut_ptr(), buf.len()) };
+        let esp_tls = unsafe { std::ptr::read_unaligned(self.raw) };
+        let read_func = esp_tls.read.unwrap();
+        let ret = unsafe { read_func(self.raw, buf.as_mut_ptr() as *mut i8, buf.len()) };
+        // ESP docs treat 0 as error, but in Rust it's common to return 0 from `Read::read` to indicate eof
```

Comment:
But then you should check for `buf.is_empty()` at the beginning of the function and return Ok(0).

Reply:
Ah right, there are two cases:

1. The buffer provided is empty
2. The connection was closed (thus eof)